ipm_library/ipm_library/utils.py

```diff
@@ -117,15 +117,16 @@ def get_field_intersection_for_pixels(
     binning_y = max(camera_info.binning_y, 1) / scale
 
     # Create rays
-    points[:, 0] = (points[:, 0] - (camera_projection_matrix[2] /
-                    binning_x)) / (camera_projection_matrix[0] / binning_x)
-    points[:, 1] = (points[:, 1] - (camera_projection_matrix[5] /
-                    binning_y)) / (camera_projection_matrix[4] / binning_y)
-    points[:, 2] = 1
+    ray_directions = np.zeros((points.shape[0], 3))
```

Comment:
Did you benchmark this for large point clouds? I am not sure if there is a better way than initing a large chunk of memory with ones and then overwriting most of it. Appending an ones array could be a bit faster Idk.

Reply:
Also the previous operation was inplace

Reply:
In addition to this it would be way faster to use np.ones instead of np.zeros and remove the `ray_directions[:, 2] = 1` line.

Reply:
My quick and dirty benchmarking showed that the new version is about 5x slower. But it needs to be said that the the "padding" is done by the user otherwise, which is not ideal and possibly moves the runtime issue there.

Reply:
It is also not that fast, but we could use the numpy pad function on the original array `np.pad(inp, ((0,0),(0,1)), constant_values=1)`

Reply:
@Flova thanks for reviewing! I wasn't familiar enough with numpy to notice that this change makes a large difference in performance, let me think about it. In terms of reading code, I'm not a fan of reusing variables if they change meaning.

Reply:
> My quick and dirty benchmarking showed that the new version is about 5x slower. But it needs to be said that the the "padding" is done by the user otherwise, which is not ideal and possibly moves the runtime issue there.

Reply:
By adding the following code (on rolling) and vice-versa on this branch,:

```diff
+   start = time.time()

    # Create rays
    points[:, 0] = (points[:, 0] - (camera_projection_matrix[2] /
                    binning_x)) / (camera_projection_matrix[0] / binning_x)
    points[:, 1] = (points[:, 1] - (camera_projection_matrix[5] /
                    binning_y)) / (camera_projection_matrix[4] / binning_y)
    points[:, 2] = 1

+   end = time.time()
+   print("old method: ", end - start)
```

I get:

old method:  2.0742416381835938e-05
new method:  1.4781951904296875e-05

which seem to indicate they take similar times.

Which part exactly is slower?

Reply:
Try scaling up the input point cloud that the operation takes a few ms. For small data, the differences are should be not significant.

Reply:
I am fine with changing the variable name while keeping the reference.

Reply:
I also did some benchmarking:

```python
inp_2d = np.ones((100000, 2))
inp_3d = np.ones((100000, 3))

def test_old(b):
     b[:, 0] = (b[:, 0] - 5)/3
     b[:, 1] = (b[:, 1] + 1)/3
     b[:, 2] = 1
     return b

def test_new(b):
     c = np.zeros((b.shape[0], 3))
     c[:, 0] = (b[:, 0] - 5)/3
     c[:, 1] = (b[:, 1] + 1)/3
     c[:, 2] = 1
     return c

In [23]: %time _ = [test_new(inp_2d) for i in range(300)]
CPU times: user 358 ms, sys: 123 ms, total: 481 ms
Wall time: 483 ms

In [24]: %time _ = [test_old(inp_3d) for i in range(300)]
CPU times: user 274 ms, sys: 53.6 ms, total: 328 ms
Wall time: 328 ms
```

There is some slow down with the new version, but the 2d to 3d would happen anyway most of the time on the user side, so I am fine with this change.
When benchmarking np.ones and np.zeros alone it is surprisingly 1000x slower than np.zeros so doing it your implementation seems nice.